lock/etcd_test.go

```diff
@@ -16,132 +16,226 @@ package lock
 
 import (
 	"errors"
-	"reflect"
 	"testing"
 
-	"go.etcd.io/etcd/client"
+	pb "go.etcd.io/etcd/api/v3/mvccpb"
+	client "go.etcd.io/etcd/client/v3"
+
 	"golang.org/x/net/context"
 )
 
+const testGroup = ""
+
+// testEtcdClient is the struct used to mock the `etcd`
+// client
 type testEtcdClient struct {
-	err  error
-	resp *client.Response
+	err     error
+	getResp *client.GetResponse
+	txn     testTxn
 }
 
-func (t *testEtcdClient) Get(ctx context.Context, key string, opts *client.GetOptions) (*client.Response, error) {
-	return t.resp, t.err
+// testTxn implements the client.Txn interface
+type testTxn struct {
+	err        error
+	txnSuccess bool
 }
 
-func (t *testEtcdClient) Set(ctx context.Context, key, value string, opts *client.SetOptions) (*client.Response, error) {
-	return t.resp, t.err
+func (t testTxn) If(cs ...client.Cmp) client.Txn {
+	return t
 }
 
-func (t *testEtcdClient) Create(ctx context.Context, key, value string) (*client.Response, error) {
-	return t.resp, t.err
+func (t testTxn) Then(ops ...client.Op) client.Txn {
+	return t
 }
 
-func TestEtcdLockClientInit(t *testing.T) {
-	for i, tt := range []struct {
-		ee      error
-		want    bool
-		group   string
-		keypath string
-	}{
-		{nil, false, "", SemaphorePrefix},
-		{client.Error{Code: client.ErrorCodeNodeExist}, false, "", SemaphorePrefix},
-		{client.Error{Code: client.ErrorCodeKeyNotFound}, true, "", SemaphorePrefix},
-		{errors.New("some random error"), true, "", SemaphorePrefix},
-		{client.Error{Code: client.ErrorCodeKeyNotFound}, true, "database", "coreos.com/updateengine/rebootlock/groups/database/semaphore"},
-		{nil, false, "prod/database", "coreos.com/updateengine/rebootlock/groups/prod%2Fdatabase/semaphore"},
-	} {
-		elc, got := NewEtcdLockClient(&testEtcdClient{err: tt.ee}, tt.group)
-		if (got != nil) != tt.want {
-			t.Errorf("case %d: unexpected error state initializing Client: got %v", i, got)
-			continue
-		}
-
-		if got != nil {
-			continue
-		}
-
-		if elc.keypath != tt.keypath {
-			t.Errorf("case %d: unexpected etcd key path: got %v want %v", i, elc.keypath, tt.keypath)
-		}
-	}
+func (t testTxn) Else(ops ...client.Op) client.Txn {
+	return t
+}
+
+func (t testTxn) Commit() (*client.TxnResponse, error) {
+	return &client.TxnResponse{Succeeded: t.txnSuccess}, t.err
+}
+
+func (t *testEtcdClient) Get(ctx context.Context, key string, opts ...client.OpOption) (*client.GetResponse, error) {
+	return t.getResp, t.err
 }
 
-func makeResponse(idx int, val string) *client.Response {
-	return &client.Response{
-		Node: &client.Node{
-			Value:         val,
-			ModifiedIndex: uint64(idx),
+func (t *testEtcdClient) Txn(ctx context.Context) client.Txn {
+	return t.txn
+}
+
+func TestEtcdLockClientInit(t *testing.T) {
+	t.Run("Success", func(t *testing.T) {
+		elc, err := NewEtcdLockClient(&testEtcdClient{
+			err:     nil,
+			getResp: &client.GetResponse{},
 		},
-	}
+			testGroup,
+		)
+		if err != nil {
+			t.Fatalf("error should be nil, got: %v", err)
```

Comment:
Similar to errors, it's good to explain what went wrong when you fail the test. For example:
```suggestion
			t.Fatalf("Failed creating new etcd lock client: %v", err)
```